### PR TITLE
Localization: optimize LRUCache

### PR DESCRIPTION
**Pull request by lfcarol**

### CODE REVIEW

File: modules/common/util/lru_cache.h [hunk 1]
```diff
@@ -169,12 +173,29 @@ class LRUCache {
...
+    return true;
+  }
+
  private:
   static constexpr size_t kDefaultCapacity = 10;
 
   const size_t capacity_;
   size_t size_;
-  std::unordered_map<K, Node<K, V>> map_;
+  std::map<K, Node<K, V>> map_;
```
Comment (storypku) [inline modules/common/util/lru_cache.h:198]: For what considerations should we change `std::unordered_map` to `std::map` ?

Comment (lfcarol) [reply modules/common/util/lru_cache.h:198]: If use unordered_map, the key which isn't built-in type must provide a hash function to make the unique key, it is not convenient for us.

File: modules/localization/msf/common/util/base_map_cache_test.cc [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+class NodeData {
+ public:
+  NodeData() : is_reserved_(false), name_("") {}
+  explicit NodeData(std::string name) : is_reserved_(false), name_(name) {}
+
+ public:
+  void SetIsReserved(bool b) { is_reserved_ = b; }
+  bool GetIsReserved() { return is_reserved_; }
+
+  void SetName(std::string name) { name_ = name; }
```
Comment (storypku) [inline modules/localization/msf/common/util/base_map_cache_test.cc:55]: `const std::string& name` ?

File: modules/localization/msf/common/util/BUILD [hunk 3]
```diff
@@ -66,6 +66,16 @@ cc_library(
     ],
 )
 
+cc_library(
+    name = "localization_msf_common_util_map_cache",
```
[the remark this answers is not in the document]
Comment (lfcarol) [reply modules/localization/msf/common/util/BUILD:70]: > Just name the target `base_map_cache`.

This is the localization module naming convention, and it should commit another pr to fix it.

File: modules/localization/msf/common/util/base_map_cache_test.cc (hunk 2 above)
Comment (lfcarol) [reply modules/localization/msf/common/util/base_map_cache_test.cc:55]: > `const std::string& name` ?

OK

File: modules/localization/msf/common/util/base_map_cache.h [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+namespace msf {
+
+template <class Key, class Element>
+using LRUCache = ::apollo::common::util::LRUCache<Key, Element*>;
+
+/**@brief The data structure of the LRUCache. */
+template <class Key, class Element, class MapLRUCache = LRUCache<Key, Element>>
+class MapNodeCache {
+ public:
+  typedef std::function<bool(Element*)> DestroyFunc;
```
Comment (changsh726) [inline modules/localization/msf/common/util/base_map_cache.h:35]: prefer `using` to `typedef`

Comment (lfcarol) [reply modules/localization/msf/common/util/base_map_cache.h]: > prefer `using` to `typedef`

OK

File: modules/localization/msf/common/util/base_map_cache_test.cc (hunk 2 above)
Comment (xiaoxq) [reply modules/localization/msf/common/util/base_map_cache_test.cc]: Actually "string_view" is the new recommendation. But it's OK to stick to `const string&` for a while.

File: modules/localization/msf/common/util/BUILD (hunk 3 above)
[the remark this answers is not in the document]
Comment (xiaoxq) [reply modules/localization/msf/common/util/BUILD:70]: Yep. Let's follow the convention in this PR and maybe fix the naming pattern in another PR.

**Comment by xiaoxq** [comment]
LGTM. Please fix the conflict.

**Comment by lfcarol** [comment]
> LGTM. Please fix the conflict.

OK

**Pull request merged**